ykrt/src/trace/swt/patch.rs

```diff
@@ -0,0 +1,143 @@
+//! This module provides functionality for dynamically patching and
+//! restoring `yk_trace_basicblock` function at runtime. It includes
+//! mechanisms to save the original instructions of a target function,
+//! patch the target function with new instructions, and restore the
+//! original instructions.
+//!
+//! This is particularly useful in scenarios such as software tracing,
+//! where the tracing function may be dynamically modified to have
+//! single return instruction and later restored to their original state
+//! for performance gain.
+//!
+//! The module relies on low-level operations like memory protection
+//! changes to manipulate function code safely. It is designed to be
+//! used with `tracer_swt` configurations, enabling conditional
+//! compilation for tracing functionalities.
+//!
+//! # Warning
+//!
+//! This module performs low-level memory operations and modifies the
+//! execution flow of functions at runtime. Improper use can lead to
+//! undefined behaviour, memory corruption, or crashes.
+
+use libc::{mprotect, size_t, sysconf, PROT_EXEC, PROT_READ, PROT_WRITE};
+use std::mem;
+use std::{ffi::c_void, sync::Once};
+
+use crate::trace::swt::yk_trace_basicblock;
+
+// This is used to ensure that the original instructions are only saved once.
+static ORIGINAL_INSTRUCTIONS_INIT: Once = Once::new();
+
+// Original instructions of the function that is patched with `PATCH_X86_INSTRUCTIONS`.
+static mut ORIGINAL_INSTRUCTIONS: [u8; 1] = [0; 1];
+
+// 0xC3 is a `ret` instruction on x86_64.
+static mut PATCH_X86_INSTRUCTIONS: [u8; 1] = [0xC3];
+
+/// This function is used to save the original instructions of a function to .
+///
+/// # Arguments
+///
+/// * `function_ptr` - A usize representing the memory address of the function.
+/// * `instructions` - A mutable pointer to a u8 where the original instructions will be saved.
+/// * `num_of_instructions` - A usize indicating the number of instructions to save.
+///
+unsafe fn save_original_instructions(
+    function_ptr: usize,
+    instructions: *mut u8,
+    num_of_instructions: usize,
+) {
+    let func_ptr: *const () = function_ptr as *const ();
+    std::ptr::copy_nonoverlapping(func_ptr as *const u8, instructions, num_of_instructions);
+}
+
+/// This function is used to patch a function instructions at runtime.
+///
+/// # Arguments
+///
+/// * `function_ptr` - A usize representing the memory address of the function to be patched.
+/// * `code` - A constant pointer to a u8 vector where the new instructions are located.
+/// * `size` - A size_t indicating the number of bytes to copy from `code`.
+///
+unsafe fn patch_function(function_ptr: usize, code: *const u8, size: size_t) {
+    let page_size = sysconf(libc::_SC_PAGESIZE) as usize;
+
+    let func_address = ((function_ptr as usize) & !(page_size - 1)) as *mut c_void;
+    let page_size_aligned = (((function_ptr as usize) + mem::size_of_val(&function_ptr))
```

Comment:
We can use `std::alloc::Layout::align` and friends to do this calculation for us.

I'm also not sure what the difference between `function_ptr` and `func_address` is.

Reply:
>I'm also not sure what the difference between function_ptr and func_address is. 

My understanding/intention here is that `function_ptr` is just the pointer to the function while `func_address` is the page address (maybe it should be renamed)

Reply:
updated :point_right: 8e00241271862bd3afdccf0f4c308217938f5098

Reply:
This might get lost but there are some comments on https://github.com/ykjit/yk/commit/8e00241271862bd3afdccf0f4c308217938f5098.

Reply:
oh I did missed that! sorry